resotocore/tools/render_dot.py

```diff
@@ -0,0 +1,334 @@
+    icon_map: Mapping[ResourceDescription, str],
+    color: int,
+) -> str:
+    return f""""{resource.uid}" [shape=plain, label=<<TABLE STYLE="ROUNDED" COLOR="{color}" BORDER="3" CELLBORDER="1" CELLPADDING="5">
+    <TR>
+        <TD SIDES="B">
+        <TABLE CELLPADDING="1" BORDER="0" CELLSPACING="0">
+        <TR>
+            <TD ALIGN="right">{render_img_tag(icon_map.get(resource.kind))}</TD>
+            <TD ALIGN="left">{resource.kind_name}</TD>
+        </TR>
+        </TABLE>
+        </TD>
+    </TR>
+    <TR>
+        <TD SIDES="B">{resource.id}</TD>
+    </TR>
+    <TR>
+        <TD BORDER="0">{resource.name}</TD>
+    </TR>
+</TABLE>>];"""
+
+
+def render_dot_header(node: str, edge: str) -> str:
+    return f"""digraph {{
+rankdir=LR
+overlap=false
+splines=true
+{node}
+{edge}"""
+
+
+def render_dot(gen: Iterator[JsonElement]) -> Generator[str, None, None]:
+    # We use the paired12 color scheme: https://graphviz.org/doc/info/colors.html with color names as 1-12
+    cit = count_iterator()
+    icon_map = generate_icon_map()
+    colors: Dict[str, int] = defaultdict(lambda: (next(cit) % 12) + 1)
+    node = "node [shape=plain colorscheme=paired12]"
+    edge = "edge [arrowsize=0.5]"
+    yield render_dot_header(node, edge)
+    in_account: Dict[str, List[str]] = defaultdict(list)
+    for item in gen:
+        if isinstance(item, dict):
+            type_name = item.get("type")
+            if type_name == "node":
+                uid = value_in_path(item, NodePath.node_id)
+                if uid:
+                    name = value_in_path_get(item, NodePath.reported_name, "n/a")
+                    kind = value_in_path_get(item, NodePath.reported_kind, "n/a")
+                    account = value_in_path_get(
+                        item, NodePath.ancestor_account_name, "graph_root"
+                    )
+                    id = value_in_path_get(item, NodePath.reported_id, "n/a")
+                    parsed_kind = parse_kind(kind)
+                    paired12 = kind_colors.get(parsed_kind, colors[kind])
+                    in_account[account].append(uid)
+                    resource = ResourceDescription(uid, name, id, parsed_kind, kind)
+                    yield render_resource(resource, icon_map, paired12)
+            elif type_name == "edge":
+                from_node = value_in_path(item, NodePath.from_node)
+                to_node = value_in_path(item, NodePath.to_node)
+                if from_node and to_node:
+                    yield f' "{from_node}" -> "{to_node}" '
+        else:
+            raise AttributeError(f"Expect json object but got: {type(item)}: {item}")
+    # All elements in the same account are rendered as dedicated subgraph
+    for account, uids in in_account.items():
+        yield f' subgraph "{account}" {{'
+        for uid in uids:
+            yield f'    "{uid}"'
+        yield " }"
+
+    yield "}"
+
+
+def ensure_assets():
+    if not os.path.exists("Assets"):
```

Comment:
how about `~/.resoto/cache/assets/`? That way it doesn't pollute the user's working directory and is always in the same directory, not relative to where the user currently is.

Reply:
Even if the directory exists we should check that it contains what we expect it to contain.

Reply:
hmm, that's a bit more complicated. I will create a directory with a more unique name for now.